### PR TITLE
Initial set of modules

### PR DESCRIPTION
**Pull request by bharathkkb**
Added modules for

- GKE Runners
- MIG Container Runners
- MIG Native Runners

Added examples for

- Simple examples on both GKE and MIG Container
- DIND runner examples on both GKE and MIG Container
- Packer/Startup script example for MIG Native Runners

### CODE REVIEW

**Comment by bharathkkb** [comment]
@craigdbarber @sethvargo PTAL

**Comment by craigdbarber** [comment]
Great work! I'm adding @danawillow to this PR to solicit her feedback. She's the TLM for the gcp-terraform provider and would have a larger context to bring to bear.

File: README.md [hunk 1]
```diff
@@ -1,43 +1,48 @@
...
 
-This module was generated from [terraform-google-module-template](https://github.com/terraform-google-modules/terraform-google-module-template/), which by default generates a module that simply creates a GCS bucket. As the module develops, this README should be updated.
+Using these Terraform module you can quickly deploy Self Hosted Github Runners for jobs in your GitHub Actions workflows
 
-The resources/services/activations/deletions that this module will create/trigger are:
+## [Self Hosted Runners on Managed Instance Groups](modules/gh-runner-mig/README.md)
 
-- Create a GCS bucket with the provided name
+The gh-runner-mig module provisions the resources required to deploy Self Hosted Runners on GCP infrastructure using Managed Instance Groups
+This includes
```
Comment (craigdbarber) [inline README.md:8]: Is this meant to be on a newline? It's being rendered on the previous line.

File: examples/gh-runner-gke-dind/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+ARG GH_RUNNER_VERSION="2.165.2"
+WORKDIR /runner
+RUN curl -o actions.tar.gz --location "https://github.com/actions/runner/releases/download/v${GH_RUNNER_VERSION}/actions-runner-linux-x64-${GH_RUNNER_VERSION}.tar.gz" && \
+    tar -zxf actions.tar.gz && \
+    rm -f actions.tar.gz && \
+    ./bin/installdependencies.sh
+
+
+WORKDIR /runner
```
Comment (craigdbarber) [inline examples/gh-runner-gke-dind/Dockerfile:40]: There seems to be a lot of logic baked into this dockerfile. Consider splitting some of it out into scripts for better readability. For inspiration, take a look at this approach: https://github.com/craigdbarber/self-hosted-runners-anthos/blob/master/Dockerfile

File: examples/gh-runner-gke-dind/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+```
+terraform init
+terraform plan
+terraform apply
+```
+
+- Step 3: Build and push the docker image
+
+```sh
+docker build --tag gcr.io/${PROJECT_ID}/runner:0.1 --tag gcr.io/${PROJECT_ID}/runner:latest .
```
Comment (craigdbarber) [inline examples/gh-runner-gke-dind/README.md:54]: An alternative to consider using here is `gcloud builds submit`. That would alleviate the need for the user to have to setup a local docker daemon.

File: examples/gh-runner-gke-dind/README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+- Step 4: Replace image in [sample k8s deployment manifest](./sample-manifests/deployment.yaml)
+
+```yaml
+.
+.
+containers:
+        - name: runner
+          image: gcr.io/${PROJECT_ID}/runner:latest
```
Comment (craigdbarber) [inline examples/gh-runner-gke-dind/README.md:65]: Consider using kustomize here for better automation. See: https://github.com/kubernetes-sigs/kustomize

File: examples/gh-runner-gke-dind/main.tf [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+module "runner-gke" {
+  source         = "../../modules/gh-runner-gke"
+  create_network = true
+  project_id     = var.project_id
+  repo_name      = var.repo_name
+  repo_owner     = var.repo_owner
+  repo_url       = var.repo_url
+  gh_token       = var.gh_token
+  machine_type   = "n1-highcpu-4"
```
Comment (craigdbarber) [inline examples/gh-runner-gke-dind/main.tf:25]: Why hard code these VM configs? Suggest instead using variables with these as default values.

File: examples/gh-runner-gke-dind/sample-manifests/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: runner-dind-deployment
+spec:
+  selector:
+    matchLabels:
+      app: runner
+  replicas: 2
```
Comment (craigdbarber) [inline examples/gh-runner-gke-dind/sample-manifests/deployment.yaml:23]: Suggest allowing this to be configured by user, with a default value.

File: examples/gh-runner-mig-dind/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    tar \
+    jq \
+    build-essential \
+    gnupg2 \
+    software-properties-common
+
+RUN curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add - && \
+    add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu bionic stable" && \
+    apt-get update && \
+    apt-get -y install docker-ce
```
Comment (craigdbarber) [inline examples/gh-runner-mig-dind/Dockerfile:30]: Have you verified that docker running inside a container works? I've heard from other folks having issues with this approach and having to use Tekton instead.

Comment (bharathkkb) [reply examples/gh-runner-mig-dind/Dockerfile:30]: Yup, I have tested the DIND runners for building from my personal test repo.

File: examples/gh-runner-gke-dind/README.md (hunk 3 above)
Comment (sethvargo) [reply examples/gh-runner-gke-dind/README.md:54]: +1. I'd rather take a dependency on Cloud Build here if you think that makes sense.

File: examples/gh-runner-gke-dind/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+          env:
+```
+
+- Step 5: Generate kubeconfig and apply the manifests for deployment and HorizontalPodAutoscaler
+
+```sh
+gcloud container clusters get-credentials your-cluster-name --zone=your-cluster-zone
```
Comment (sethvargo) [inline examples/gh-runner-gke-dind/README.md:72]: Consider prefixing "command to run" with `$` to separate it from "output to expect"

File: examples/gh-runner-gke-dind/main.tf [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  min_node_count = 3
```
Comment (sethvargo) [inline examples/gh-runner-gke-dind/main.tf:26]: Variables for these please

File: examples/gh-runner-gke-dind/outputs.tf [hunk 10]
```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright 2019 Google LLC
```
Comment (sethvargo) [inline examples/gh-runner-gke-dind/outputs.tf:2]: Date here and throughout

File: examples/gh-runner-gke-dind/outputs.tf [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+
+output "kubernetes_endpoint" {
+  description = "The cluster endpoint"
+  sensitive   = true
+  value       = module.runner-gke.kubernetes_endpoint
+}
+
+output "client_token" {
```
Comment (sethvargo) [inline examples/gh-runner-gke-dind/outputs.tf:23]: Why do we need to output this?

File: README.md (hunk 1 above)
Comment (bharathkkb) [reply README.md:31]: oops NL. Fixed

File: examples/gh-runner-gke-dind/Dockerfile (hunk 2 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/Dockerfile]: Done

File: examples/gh-runner-gke-dind/README.md (hunk 3 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/README.md]: Done, added cloudbuild.yamls for all examples and updated instructions.

File: examples/gh-runner-gke-dind/README.md (hunk 4 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/README.md]: added kustomize instructions and kustomization.yamls

File: examples/gh-runner-gke-dind/main.tf (hunk 5 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/main.tf]: This was something I was testing with, didn't intend to be part of the example. Just for some clarity- anything in examples/* is just an example of how to use the actual modules which reside in modules/*. I have used variables with sane defaults for the modules which will be what the end user directly uses.

File: examples/gh-runner-gke-dind/main.tf (hunk 9 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/main.tf]: This was something I was testing with, didnt intend to be part of the example. Just for some clarity- anything in examples/* is just an example of how to use the actual modules which reside in modules/*. I have used variables with sane defaults for the modules which will be what the end user directly uses.

File: examples/gh-runner-gke-dind/sample-manifests/deployment.yaml (hunk 6 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/sample-manifests/deployment.yaml:23]: Now that we are using kustomize, they can override this if needed.

File: examples/gh-runner-gke-dind/README.md (hunk 8 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/README.md]: Done

File: examples/gh-runner-gke-dind/outputs.tf (hunk 10 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/outputs.tf]: Done

File: examples/gh-runner-gke-dind/outputs.tf (hunk 11 above)
Comment (bharathkkb) [reply examples/gh-runner-gke-dind/outputs.tf:23]: I use these in some of the inspec tests here. https://github.com/terraform-google-modules/terraform-google-github-actions-runners/blob/53beb1d26ba568732745a64f0d2b6c2158e4c8d7/test/integration/gh-runner-gke/controls/kubectl.rb#L19

**Comment by bharathkkb** [comment]
Hi All,
I am merging this PR but we can discuss if there are any additional outstanding items via issues.

**Pull request merged**